src/sciline/container.py

```diff
@@ -0,0 +1,94 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+import typing
+from functools import wraps
+from typing import Callable, List, Type, TypeVar, Union
+
+import injector
+from dask.delayed import Delayed
+
+T = TypeVar('T')
+
+
+class UnsatisfiedRequirement(Exception):
+    pass
+
+
+class Container:
+    def __init__(self, inj: injector.Injector, /, *, lazy: bool) -> None:
+        self._injector = inj
+        self._lazy = lazy
+
+    def get(self, tp: Type[T], /) -> Union[T, Delayed]:
+        try:
+            # We are slightly abusing Python's type system here, by using the
+            # injector to get T, but actually it returns a Delayed that can
+            # compute T. self._injector does not know this due to how we setup the
+            # bindings. We'd like to use Delayed[T], but that is not supported yet:
+            # https://github.com/dask/dask/pull/9256
+            task: Delayed = self._injector.get(tp)  # type: ignore
+        except injector.UnsatisfiedRequirement as e:
+            raise UnsatisfiedRequirement(e) from e
+        return task if self._lazy else task.compute()
+
+
+def _delayed(func: Callable) -> Callable:
+    """
+    Decorator to make a function return a delayed object.
+
+    In contrast to dask.delayed, this uses functools.wraps, to preserve the
+    type hints, which is a prerequisite for injector to work.
```

Comment:
Do you know why dask doesn't use `wraps`? They might have a good reason.

Reply:
I suppose because it would be lie? The types are actually different.

Reply:
Do why does this code?

Reply:
See also the comment in `Container.get`: `dask.Delayed` is not a generic, so we cannot use `dask.Delayed[T]`. This implies that `injector` could not do its job.